cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,951 @@
+#include <thrust/execution_policy.h>
+#include <thrust/fill.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/optional.h>
+#include <thrust/scatter.h>
+#include <thrust/sort.h>
+#include <thrust/transform.h>
+#include <thrust/tuple.h>
+
+#include <prims/edge_bucket.cuh>
+#include <prims/extract_transform_e.cuh>
+#include <prims/fill_edge_property.cuh>
+#include <prims/reduce_op.cuh>
+#include <prims/transform_e.cuh>
+#include <prims/transform_reduce_dst_nbr_intersection_of_e_endpoints_by_v.cuh>
+#include <prims/transform_reduce_e.cuh>
+#include <prims/update_edge_src_dst_property.cuh>
+
+namespace cugraph {
+
+// FIXME : This will be deleted once edge_triangle_count becomes public
+template <typename vertex_t, typename edge_t, bool store_transposed, bool multi_gpu>
+rmm::device_uvector<edge_t> edge_triangle_count(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu> const& graph_view,
+  raft::device_span<vertex_t> edgelist_srcs,
+  raft::device_span<vertex_t> edgelist_dsts);
+
+template <typename vertex_t, typename edge_t, typename EdgeBuffer>
+struct unroll_edge {
+  edge_t num_valid_edges{};
+  raft::device_span<edge_t> num_triangles{};
+  EdgeBuffer edge_to_unroll_first{};
+  EdgeBuffer transposed_valid_edge_first{};
+  EdgeBuffer transposed_valid_edge_last{};
+  EdgeBuffer transposed_invalid_edge_last{};
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t i) const
+  {
+    // edges are sorted with destination as key so reverse the edge when looking it
+    auto pair = thrust::make_tuple(thrust::get<1>(*(edge_to_unroll_first + i)),
+                                   thrust::get<0>(*(edge_to_unroll_first + i)));
+    // Find its position in either partition of the transposed edgelist
+    // An edge can be in found in either of the two partitions (valid or invalid)
+
+    auto itr = thrust::lower_bound(
+      thrust::seq, transposed_valid_edge_last, transposed_invalid_edge_last, pair);
+    auto idx = thrust::distance(transposed_valid_edge_last, itr) + num_valid_edges;
+
+    if (*itr != pair) {
+      // The edge must be in the first boundary
+      itr = thrust::lower_bound(
+        thrust::seq, transposed_valid_edge_first, transposed_valid_edge_last, pair);
+      idx = thrust::distance(transposed_valid_edge_first, itr);
+    }
+
+    assert(*itr == pair);
+    cuda::atomic_ref<edge_t, cuda::thread_scope_device> atomic_counter(num_triangles[idx]);
+    auto r = atomic_counter.fetch_sub(edge_t{1}, cuda::std::memory_order_relaxed);
+  }
+};
+
+template <typename vertex_t>
+rmm::device_uvector<vertex_t> prefix_sum_valid_and_invalid_edges(
+  // The edgelist is segmented into 2 partitions (valid and invalid edges)
+  // and edges to be unrolled can be either in the valid or the invalid edge
+  // partition.
+  raft::handle_t const& handle,
+  raft::device_span<vertex_t const> invalid_dsts,
+  raft::device_span<vertex_t const> edgelist_dsts)
```

Comment:
This function is something much more generic than just computing prefix_sum of valid and invalid edges.

We can better rename to reflect that (and in the future, we may re-locate this function as a general utility function for graph algorithm implementations).

May better rename to something like

```
compute_prefix_sum(
  raft::handle_t const& handle,
  raft::device_span<vertex_t const> sorted_vertices, <= previously, edgelist_dsts
  raft::device_span<vertex_t const> query_vertices); <= previously, invalid_dsts
```

Reply:
And this is just a suggestion, but following the CSR convention, I use `offsets` for prefix sum arrays having the size of `num_query_elemetns` +1 and `displacements` for prefix sum arrays having the size of `num_query_elements`. You may consider following this naming convention.